File: tcpduplex/fingerprint.go
package tcpduplex

import (
	"crypto/sha256"
)

// PeerPublicKeyFingerprint returns SHA256(raw X25519 public key bytes) for use with HandshakeAuth.ExpectedPeerPubKeySHA256.
func PeerPublicKeyFingerprint(pubKey []byte) [sha256.Size]byte {
	return sha256.Sum256(pubKey)
}

File: lib/src/main/kotlin/com/hdmain/tcpduplex/Fingerprint.kt
package com.hdmain.tcpduplex

import java.security.MessageDigest

/** Returns SHA-256(raw X25519 public key bytes) for [HandshakeAuth.expectedPeerPublicKeySha256]. */
fun peerPublicKeyFingerprint(publicKey: ByteArray): ByteArray =
    MessageDigest.getInstance("SHA-256").digest(publicKey)

/** Returns a lowercase hex SHA-256 fingerprint suitable for logging and config files. */
fun peerPublicKeyFingerprintHex(publicKey: ByteArray): String =
    peerPublicKeyFingerprint(publicKey).joinToString("") { "%02x".format(it) }
